### DISCUSSION TITLE
Min.io upload is slower than file system

### DISCUSSION THREAD

**Discussion by arazmakhnin** [minio/minio#17116] [Q&A] [answered]
Hi. I installed one node min.io server to test it's features and performance. But unexpectedly for me it is really slow: even slower than just a regular file copy on my hard drive. 
How I test (for example, for 512 Mb file):
- run `cp test512m.dat /opt/data/min.io` takes ~1200 ms
- run `dd if=/dev/zero of=/opt/data/min.io/test512m.dat bs=512M count=1 oflag=direct` takes same 1200 ms with write speed ~480 Mbps
- run `mcli cp test512m.dat myminio/testbucket` takes 4 seconds with speed 130Mbps

I expected that Min.io should be much faster than file system as one of it benefit. Is it correct? And should I make some changes in configuration to speed up upload?

Linux: Ubuntu 22.04.1 LTS
Min.io: 2023-04-20T17-56-55Z

**Comment by klauspost** [comment]
MinIO writes to your file system. How should it be faster than the file system?

`mc cp` does more than just write files. Performance testing can be done with [warp](https://github.com/minio/warp).

**Comment by arazmakhnin** [comment]
Well, there are lots of articles saying this. For example, [here](https://medium.com/@tiwari_nitish/minio-benchmarks-with-cosbench-81704a8f0178). Hardware: 1x2400 NVMe disk with "uncached disk write speed 720 MiB/s". They got result speed on big files around 2.5 GiB/s.

Another example from [min.io resourses](https://resources.min.io/l/library?overlay_url=https%3A%2F%2Fresources.min.io%2Fc%2Fminio-throughput-benchmark-on-hard-disk-drives%3Fx%3DP9k0ng%26lx%3DexvNTw): 24 disks with 137 MiB/s write speed each. Got result 9.4 Gb/s put speed.

For me, if min.io would be comparable with file system, it would be fine. But it's 3-4 times slower. So, trying to understand what's wrong.

**Reply by klauspost** [reply]
@arazmakhnin Writes are distributed across disks, leading to bigger aggregate performance.

**Reply by arazmakhnin** [reply]
@klauspost So, on one node setup I can't get performance that is comparable with plain file copy, right?

**Reply by klauspost** [reply]
@arazmakhnin Cached copies, no.

**Reply by arazmakhnin** [reply]
Not sure I got what is cached copies. It's a part of min.io?

**Comment by harshavardhana** [comment]
> * run `cp test512m.dat /opt/data/min.io` takes ~1200 ms

This is cached @arazmakhnin

**Comment by harshavardhana** [comment, marked as answer]
> * run `dd if=/dev/zero of=/opt/data/min.io/test512m.dat bs=512M count=1 oflag=direct` takes same 1200 ms with write speed ~480 Mbps

This is also wrong test bs=512M is meaningless no one can buffer up 512M worth of streams use something like 64k or 128k 

```
dd if=/dev/zero of=/opt/data/min.io/test512m.dat bs=128k count=4096 oflag=direct conv=fdatasync
```
